### DISCUSSION TITLE
How to use queryParam helper inside data helper with JSONPath syntax

### DISCUSSION THREAD

**Discussion by jcbenck** [mockoon/mockoon#1194] [Help] [answered]
I am using Mockoon to create a dynamic response that depends on the queryParam perPage. I have a databucket called v7es that contains an array of objects. I want to use the data template helper to get the data from the databucket up to a specific index that is determined by the queryParam perPage. For example, if the queryParam perPage is 10, I want to get the first 10 elements of the array. I have tried to use the JSONPath syntax with the data helper, like this:

`{{ data 'v7es' '$[0:10]' }}
`
This works fine, but I want to make the 10 value dynamic, so I tried to use the queryParam helper inside the data helper, like this:

`{{ data 'v7es' '$[0:(queryParam 'perPage')]' }}
`
But this gives me a parse error:

> Error while serving the content: Parse error on line 8: …’ ‘$[0:(queryParam ‘perPage’)]’ }}, -----------------------^ Expecting ‘CLOSE_RAW_BLOCK’, ‘CLOSE’, ‘CLOSE_UNESCAPED’, ‘OPEN_SEXPR’, ‘CLOSE_SEXPR’, ‘ID’, ‘OPEN_BLOCK_PARAMS’, ‘STRING’, ‘NUMBER’, ‘BOOLEAN’, ‘UNDEFINED’, ‘NULL’, ‘DATA’, got ‘INVALID’

I don’t understand why this is not working, and I can’t find any documentation or examples on how to use the queryParam helper inside the data helper. Is there a way to make the JSONPath dynamic with the data helper? How can I fix this parse error?

**Comment by jcbenck** [comment]
Maybe this can be related to my question but I noticed that
`{{concat '1:' (queryParam 'perPage')}},`
results in: 1: 10,

`{{concat '1-' (queryParam 'perPage')}},`
results in: 1-10,

`{{concat '1' 2: (queryParam 'perPage')}}`
results in: 110,

what is the tricky about the `:` ?

**Reply by 255kb** [reply]
I'm not sure how the "2:" gets interpreted (apparently ignored), but you are supposed to pass strings to concat or eventually numbers/booleans.

**Comment by 255kb** [comment, marked as answer]
First, you have single quotes in single quotes in your JSONPath string, which breaks the compiler. 
But even by solving this it wouldn't work as the JSONPath string will no get parsed for helpers. You need to use [`concat`](https://mockoon.com/docs/latest/templating/mockoon-helpers/#concat): `{{ data 'bucketid' (concat '$[0:' (queryParam 'perPage') ']') }}`.

**Comment by jcbenck** [comment]
Thank you @255kb, it works! btw is there a way I can define a variable and a default value for it if the original value is null?
